ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/TopicStream.java

```diff
@@ -82,116 +87,144 @@ public RecordFormatter(SchemaRegistryClient schemaRegistryClient, String topicNa
     }
 
     public Format getFormat() {
-      return format;
+      return formatter == null ? Format.UNDEFINED : formatter.getFormat();
+    }
+
+    private Formatter getFormatter(final ConsumerRecord<String, Bytes> record) {
+      return Arrays.stream(Format.values())
+          .map(f -> f.maybeGetFormatter(topicName, record, avroDeserializer, dateFormat))
+          .filter(Optional::isPresent)
+          .map(Optional::get)
+          .findFirst()
+          .orElseThrow(() -> new RuntimeException("Unexpected"));
```

Comment:
"Unexpected" seems like a terse message. Maybe "Could not find an appropriate formatter"?

Reply:
It should never happen. The code always defaults to String.